src/core/mac/mac.cpp

```diff
@@ -569,6 +596,12 @@ void Mac::ScheduleNextTransmission(void)
     else
     {
         mState = kStateIdle;
+
+        if (mDisableRadioOnIdle)
+        {
+            mDisableRadioOnIdle = false;
+            DisableRadio();
+        }
```

Comment:
Here is a suggested solution (without a counter):
- Remove the `mDisableRadioOnIdle`
- Before starting scan, check if `mRadioOn` is false, directly call `otPlatRadioEnable(mNetif.GetInstance())` to enable radio for scanning.
- At the end of scan, check the status of `mRadioOn` if it's not true,  directly call `otPlatRadioDisable()`
- In `EnableRadio()` check the state, if we are in in scan state, just set the flag `mRadioOn` (since radio is already enabled by the scan operation and then won't be disabled at the end of scan), otherwise we enable the radio and set the `mRadioOn` to true. We can do similar pattern in `DisableRadio()`.